### PR TITLE
refactor discrete curves

### PR DESCRIPTION
**Pull request by SaitejaUtpala**
Refator discrete Curves

### CODE REVIEW

File: geomstats/learning/pca.py [hunk 1]
```diff
@@ -286,6 +286,7 @@ def _fit(self, X, base_point=None):
                 )
 
         # Center data - the mean should be 0 if base_point is the Frechet mean
+        print("hape after", X.shape)
```
Comment (shubhamtalbar96) [inline geomstats/learning/pca.py:289]: Typo with "hape".

Comment (SaitejaUtpala) [reply geomstats/learning/pca.py]: Thanks! Actually this is stray code.  I removed it completely

File: tests/tests_geomstats/test_discrete_curves.py [hunk 2]
```diff
@@ -1,315 +1,617 @@
...
+    skip_test_exp_geodesic_ivp = True
+    skip_test_parallel_transport_ivp_is_isometry = True
+    skip_test_parallel_transport_bvp_is_isometry = True
+    skip_test_dist_is_norm_of_log = tf_backend()
+    skip_test_dist_is_symmetric = tf_backend()
+    skip_test_squared_dist_is_symmetric = tf_backend()
+    skip_test_inner_product_is_symmetric = tf_backend()
 
-        self.assertAllClose(gs.shape(result), (n_samples,))
+    class L2CurvesMetricTestData(_RiemannianMetricTestData):
```
Comment (ninamiolane) [inline tests/tests_geomstats/test_discrete_curves.py:117]: As per @lpereira95 refactoring, we do not have nested testdata classes anymore. This should go in: https://github.com/geomstats/geomstats/tree/master/tests/data

File: tests/tests_geomstats/test_discrete_curves.py [hunk 3]
```diff
@@ -618,63 +827,511 @@ def test_space_derivative(self):
...
+        expected = proj
+        result = planar_closed_curve.project(proj)
+        self.assertAllClose(result, expected)
+
+        result = proj[-1, :]
+        expected = proj[0, :]
+        self.assertAllClose(result, expected, rtol=10 * gs.rtol)
+
+
+class TestClosedSRVMetric(RiemannianMetricTestCase, metaclass=Parametrizer):
```
Comment (ninamiolane) [inline tests/tests_geomstats/test_discrete_curves.py:952]: Nothing works? Were these tested in the current version of our tests, or are these new tests that happen to not run?

**Review by ninamiolane — CHANGES_REQUESTED** [review]
The TestData structures should be put in : https://github.com/geomstats/geomstats/tree/master/tests/data.

What is going on with the tests being all skipped in ClosedCurves? Are these new tests?

**Comment by SaitejaUtpala** [comment]
> The TestData structures should be put in : https://github.com/geomstats/geomstats/tree/master/tests/data.
> 
> What is going on with the tests being all skipped in ClosedCurves? Are these new tests?

Nice Catch! I forgot. Let me adapt.

File: tests/tests_geomstats/test_discrete_curves.py (hunk 3 above)
Comment (SaitejaUtpala) [reply tests/tests_geomstats/test_discrete_curves.py:952]: yup.. this was not tested at all before. These are new tests.

File: tests/data/discrete_curves_data.py [hunk 4]
```diff
@@ -0,0 +1,637 @@
...
+        )
+
+    def log_after_exp_test_data(self):
+        return self._log_after_exp_test_data(
+            self.metric_args_list,
+            self.space_list,
+            self.shape_list,
+            self.n_tangent_vecs_list,
+            rtol=gs.rtol * 100,
+            atol=gs.atol * 10000,
```
Comment (ninamiolane) [inline tests/data/discrete_curves_data.py:164]: Do you know what is going on with exp and log here?

For future work: the tangent vectors generated to test log_after_exp should be generated with their norm being less than the injectivity radius of the metric defining the exp. 

Nicolas put the injectivity radii of many metrics, e.g. here for the hypersphere:
https://github.com/geomstats/geomstats/blob/bc82fbfb5aae9dbb44e2bd99ddfd5fe9f6a427ed/geomstats/geometry/hypersphere.py#L1088

Thus, log_after_exp_test_data could be adapted in the parent class.

**Review by ninamiolane — APPROVED** [review]
Excellent, thanks!

**Pull request merged**